src/benchmark/Akka.Benchmarks/Dispatch/DispatcherBenchmarks.cs

```diff
@@ -0,0 +1,121 @@
+// // </copyright>
+// //-----------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Akka.Actor;
+using Akka.Benchmarks.Configurations;
+using Akka.Configuration;
+using Akka.Dispatch;
+using BenchmarkDotNet.Attributes;
+using IRunnable = Akka.Dispatch.IRunnable;
+
+namespace Akka.Benchmarks.Dispatch
+{
+    /// <summary>
+    /// Used to test the performance of various dispatchers
+    /// </summary>
+    [Config(typeof(MicroBenchmarkConfig))]
+    public class DispatcherBenchmarks
+    {
+        private ActorSystem _sys;
+        private DefaultDispatcherPrerequisites _prereqs;
+
+        private MessageDispatcher _dispatcher;
+
+        [Params(1_000_000)] // higher values will cause the CallingThreadDispatcher to stack overflow
+        public int TaskCount { get; set; }
+
+        [ParamsSource(nameof(AllConfigurators))]
+        public DispatcherConfig Configurator { get; set; }
+
+        public class DispatcherConfig
+        {
+            public DispatcherConfig(Config config, string name)
+            {
+                Config = config;
+                Name = name;
+            }
+
+            public Config Config { get; }
+
+            public string Name { get; }
+
+            public override string ToString()
+            {
+                return Name;
+            }
+        }
+
+        private static readonly Config DefaultConfig = Akka.Remote.Configuration.RemoteConfigFactory.Default()
+            .WithFallback(Akka.Configuration.ConfigurationFactory.Default());
+
+        public IEnumerable<DispatcherConfig> AllConfigurators()
+        {
+            yield return new DispatcherConfig(DefaultConfig.GetConfig("akka.actor.default-dispatcher"), "DefaultThreadPool");
+            yield return new DispatcherConfig(DefaultConfig.GetConfig("akka.actor.internal-dispatcher"), "ForkJoin(sys)");
+            yield return new DispatcherConfig(DefaultConfig.GetConfig("akka.remote.default-remote-dispatcher"), "ForkJoin(remote)");
+            yield return new DispatcherConfig(@" executor = channel-executor
+                throughput=30", "ChannelD");
+            yield return new DispatcherConfig(@" executor = task-executor
+                throughput=30", "TaskD");
+        }
+
+        private TaskCompletionSource<int> _tcs;
+
+        public sealed class RunnableTarget : IRunnable
+        {
+            private readonly TaskCompletionSource<int> _tcs;
+            private readonly int _target;
+            private int _counter = 0;
+
+            public RunnableTarget(TaskCompletionSource<int> tcs, int target)
+            {
+                _tcs = tcs;
+                _target = target;
+            }
+
+            public void Run()
+            {
+                if (Interlocked.Increment(ref _counter) >= _target)
+                {
+                    _tcs.TrySetResult(_counter);
```

Comment:
Try to force all operations to complete before allowing each benchmark iteration to exit